Proposed feature deprecation process (#2214)
sourcegraph/about: handbook/product/deprecation_process.md

# Feature deprecation process

Sometimes we need to deprecate features that we've already released. We follow this process to minimize the impact of that deprecation on both our customers and our internal teams. 

## Choosing to deprecate a feature

You should first determine the potential impact of the feature deprecation: 

> - Are any users are using the feature? Who are they? Is there another feature that also meets their needs? 

Reviewer: ```suggestion
- Are any users using the feature? Who are they? Is there another feature that also meets their needs? 
```

- Are any Sourcegraph teams using the feature, or a subset of its code?
- Are any sales prospects specifically interested in the feature? 

If the answers to these questions suggest that deprecation is the best path forward, you should next document the deprecation. 

## Documenting a feature's deprecation 

You should then create a blog post to publish when we deprecate the feature. The blog post should contain:

- Why we're deprecating the feature.
- The deprecation timeline.
- Workarounds for the feature's users, if relevant.

We document this in a blog post because blog posts are publicly accessible, timestamped, and easy to send directly to customers or prospects.

## Preparing internal teams for the deprecation

Once you have the blog post, you should share it with the CE team by sharing the blog post in #CE. This gives the CE team context so they can support customers or do high touch outreach. 

> You should also share the blog post in #dev-announce so the product and engineering teams are aware. 

Author: ```suggestion
You should also share the blog post in #general so all teams are aware of it.  
```

Answer any questions or concerns about the deprecation that teams may have. 

## Deprecating the feature

Only after sharing this deprecation internally should you do any engineering work needed to deprecate the feature.   

## Publishing the blog post 

> The blog post should be published when the first version that includes the deprecation is released. The release blog post changelog section should link to the deprecation post.  

Reviewer: Thinking about communication around deprecating features, I'm thinking if it would also be helpful to make sure we mention it in the Product updates/Progress slide we have in company meeting. Would it help to have it as part of the documented communication process or is it already implied as it's a relevant part of the changelog?

Author: I think it's worth calling out at the company meeting, and worth adding a note here – who knows if the product slide will always come straight from the changelog. I'll add a quick sentence.

Author: Actually, on second thought, I feel like it's most useful in the company meeting to gather "be aware of this + speak up if there's pertinent information you have about this feature being deprecated" and Slack is a much better place for that. If folks wonder if the deprecation has happened yet, the blog post published/not published should be a good signal. So do you think #dev-announce is the right channel to get this feedback and awareness @macraig or should it be posted in something like #general instead?

Reviewer: That's a good point! I think I'd go for #general to make sure folks from non-dev related teams are aware (not sure if everyone non-dev actively follows #dev-announce). Another option would be #product, but we'll likely have the same problem than with #dev-announce as Slack is getting crowded and people will tend to follow channels directly related to their teams.

Author: Great, I'll make that change.
